packages/ckeditor5-widget/src/widgettypearound/widgettypearound.js

```diff
@@ -0,0 +1,250 @@
+				} )
+				// ...and update widgets' classes depending on possible positions for paragraph insertion.
+				.forEach( ( { item: widgetViewElement } ) => {
+					const positions = getWidgetTypeAroundPositions( widgetViewElement );
+
+					// Remove all classes. In theory we could remove only these that will not be added a few lines later,
+					// but since there are only two... KISS.
+					writer.removeClass( POSSIBLE_INSERTION_POSITIONS.map( positionToWidgetCssClass ), widgetViewElement );
+
+					// Set CSS classes related to possible positions. They are used so the UI knows which buttons to display.
+					writer.addClass( positions.map( positionToWidgetCssClass ), widgetViewElement );
+				} );
+		} );
+	}
+
+	/**
+	 * Registers a `mousedown` listener for the view document which intercepts events
+	 * coming from the type around UI, which happens when a user clicks one of the buttons
+	 * that insert a paragraph next to a widget.
+	 *
+	 * @private
+	 */
+	_enableInsertingParagraphsOnButtonClick() {
+		const editor = this.editor;
+		const editingView = editor.editing.view;
+
+		editingView.document.on( 'mousedown', ( evt, domEventData ) => {
+			const button = getClosestTypeAroundDomButton( domEventData.domTarget );
+
+			if ( !button ) {
+				return;
+			}
+
+			const buttonPosition = getTypeAroundButtonPosition( button );
+			const widgetViewElement = getClosestWidgetViewElement( button, editingView.domConverter );
+
+			this._insertParagraph( widgetViewElement, buttonPosition );
+
+			domEventData.preventDefault();
+			evt.stop();
+		} );
+	}
+}
+
+// Injects the type around UI into a view widget instance.
+//
+// @param {module:engine/view/view} editingView
+// @param {Object.<String,String>} buttonTitles
+// @param {module:engine/view/element~Element} widgetViewElement
+function injectUIIntoWidget( editingView, buttonTitles, widgetViewElement ) {
+	editingView.change( writer => {
+		const typeAroundWrapper = writer.createUIElement( 'div', {
+			class: 'ck ck-reset_all ck-widget__type-around'
+		}, function( domDocument ) {
+			const wrapperDomElement = this.toDomElement( domDocument );
+
+			injectButtons( wrapperDomElement, buttonTitles );
+
+			return wrapperDomElement;
+		} );
+
+		// Inject the type around wrapper into the widget's wrapper.
+		writer.insert( writer.createPositionAt( widgetViewElement, 'end' ), typeAroundWrapper );
+	} );
+}
+
+// FYI: Not using the IconView class because each instance would need to be destroyed to avoid memory leaks
+// and it's pretty hard to figure out when a view (widget) is gone for good so it's cheaper to use raw
+// <svg> here.
+//
+// @param {HTMLElement} wrapperDomElement
+// @param {Object.<String,String>} buttonTitles
+function injectButtons( wrapperDomElement, buttonTitles ) {
+	// Do the SVG parsing once and then clone the result <svg> DOM element for each new
+	// button. There could be dozens of them during editor's lifetime.
+	if ( !CACHED_RETURN_ARROW_ICON ) {
```

Comment:
We don't necessarily need to delay parsing this. In a great number of cases, the icon will be needed anyway.